### DISCUSSION TITLE
TextClause for a JSON string representation or datetime object

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7391] [Usage Questions]
Hi

I have a Json representation, let's assume :

```python
my_json = json.dumps({"dummy_key":"dummy_value"})
```
When I apply an insert with a `from_select` statement like this : 
```python
    query = insert(MyTable).from_select(['attr1','attr2','attr3'], 
                                         select(column('attr1'), text(my_json), column('attr3'))
                                         .where(MyTable.id == target_id))
   session.execute(query)
```
I got this error : 

```python
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) unrecognized token: "{"
```
How should I insert litteral json value with the syntax `from_select` then ?

I found that I have the same issue with a datetime string : 

```python 
my_datetime = str(dummy_date) # Which output '2021-12-02 17:57:53.080792'
```
If I parse it to textClause
```python
query = insert(MyTable).from_select(['attr1','attr2','attr3'], 
                                         select(column('attr1'), text(my_datetime), column('attr3'))
                                         .where(MyTable.id == target_id))
 session.execute(query)
```
I will have a similar SQL syntax error : 

```python
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) near "17": syntax error
```
And this seems to be because the `text` method will send the string as a plain SQL (without adding quotes to a string value) sot my datetime will be sent as : 

```sql
SELECT .... VALUES ( 2021-12-02 17:57:53.080792 )
```

This is for sure a not correct SQL Syntaxe...how could I wourkaround this ?

**Comment by jossefaz** [comment]
Ok I've found the way to escape it :

```python

text(f"\'{my_json}\'")

```

**Reply by zzzeek** [reply]
What you have there is in fact kind of a SQL injection vulnerability, depending on the way the data formats.    you'd be better off doing it like this:

``literal(my_json, type_=JSON)``


see [literal](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=literal#sqlalchemy.sql.expression.literal)

**Reply by kiddten** [reply]
How to handle that case if I need {"entity_id": MyTable.c.id"} dynamic key as part of json?
